server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java

```diff
@@ -120,47 +127,89 @@ public void parse(ParseContext context) throws IOException {
     }
 
     private void indexShape(ParseContext context, Object luceneShape) {
-        if (luceneShape instanceof GeoPoint) {
-            GeoPoint pt = (GeoPoint) luceneShape;
-            indexFields(context, LatLonShape.createIndexableFields(name(), pt.lat(), pt.lon()));
-        } else if (luceneShape instanceof double[]) {
-            double[] pt = (double[]) luceneShape;
-            indexFields(context, LatLonShape.createIndexableFields(name(), pt[1], pt[0]));
-        } else if (luceneShape instanceof Line) {
-            indexFields(context, LatLonShape.createIndexableFields(name(), (Line)luceneShape));
-        } else if (luceneShape instanceof Polygon) {
-            indexFields(context, LatLonShape.createIndexableFields(name(), (Polygon) luceneShape));
-        } else if (luceneShape instanceof double[][]) {
-            double[][] pts = (double[][])luceneShape;
-            for (int i = 0; i < pts.length; ++i) {
-                indexFields(context, LatLonShape.createIndexableFields(name(), pts[i][1], pts[i][0]));
-            }
-        } else if (luceneShape instanceof Line[]) {
-            Line[] lines = (Line[]) luceneShape;
-            for (int i = 0; i < lines.length; ++i) {
-                indexFields(context, LatLonShape.createIndexableFields(name(), lines[i]));
-            }
-        } else if (luceneShape instanceof Polygon[]) {
-            Polygon[] polys = (Polygon[]) luceneShape;
-            for (int i = 0; i < polys.length; ++i) {
-                indexFields(context, LatLonShape.createIndexableFields(name(), polys[i]));
-            }
-        } else if (luceneShape instanceof Rectangle) {
-            // index rectangle as a polygon
-            Rectangle r = (Rectangle) luceneShape;
-            Polygon p = new Polygon(new double[]{r.minLat, r.minLat, r.maxLat, r.maxLat, r.minLat},
-                new double[]{r.minLon, r.maxLon, r.maxLon, r.minLon, r.minLon});
-            indexFields(context, LatLonShape.createIndexableFields(name(), p));
-        } else if (luceneShape instanceof Object[]) {
-            // recurse to index geometry collection
-            for (Object o : (Object[])luceneShape) {
-                indexShape(context, o);
-            }
+        if (luceneShape instanceof Geometry) {
+            ((Geometry) luceneShape).visit(new GeometryVisitor<Void>() {
```

Comment:
This is kind of hairy. Can we maybe have a `protected` member class that implements `GeometryVisitor<Void>` instead of doing this inline?

Reply:
I'm thinking it might be cleaner to do something like this

```
    private void indexShape(ParseContext context, Object luceneShape) {
        if (luceneShape instanceof Geometry) {
            ((Geometry) luceneShape).visit(new LuceneGeometryIndexer(context));
        } else {
            throw new IllegalArgumentException("invalid shape type found [" + luceneShape.getClass() + "] while indexing shape");
        }
    }

...

protected class LuceneGeometryIndexer implements GeometryVisitor<Void> {
...
```

Reply:
Good idea. I don't think it needs to be protected though. We can make it private here.